package/MDAnalysis/converters/RDKitInferring.py

```diff
@@ -0,0 +1,699 @@
+            return self.assign_from_template_with_adjusted_hydrogens(mol)
+        return AssignBondOrdersFromTemplate(self.template, mol)
+
+    def assign_from_template_with_adjusted_hydrogens(
+        self, mol: "Chem.Mol", index_field: str = "_MDAnalysis_index"
+    ) -> "Chem.Mol":
+        """Temporarily removes hydrogens from the molecule before assigning
+        bond orders and charges from the template, then adds them back.
+
+        Parameters
+        ----------
+        mol : rdkit.Chem.rdchem.Mol
+            Molecule to assign bond orders and charges to.
+        index_field : str, default = "_MDAnalysis_index"
+            Atom property corresponding to a unique integer that can used to
+            put back the hydrogen atoms that were removed before matching,
+            and sort them back to the original order. Must be accessible
+            through ``atom.GetIntProp``.
+        """
+        # remove explicit hydrogens and assign BO from template
+        mol_no_h = Chem.RemoveAllHs(mol, sanitize=False)
+        new = AssignBondOrdersFromTemplate(self.template, mol_no_h)
+        # mapping between AtomGroup index of atom bearing H --> H atom(s)
+        atoms_with_hydrogens = defaultdict(list)
+        for atom in mol.GetAtoms():
+            if atom.GetAtomicNum() == 1:
+                atoms_with_hydrogens[
+                    atom.GetNeighbors()[0].GetIntProp(index_field)
+                ].append(atom)
+        # mapping between atom that should be bearing a H in RWMol and
+        # corresponding H(s)
+        reverse_mapping = {}
+        for atom in new.GetAtoms():
+            if (idx := atom.GetIntProp(index_field)) in atoms_with_hydrogens:
+                reverse_mapping[atom.GetIdx()] = atoms_with_hydrogens[idx]
+                # Set ExplicitH to 0 to avoid valence errors when adding Hs
+                atom.SetNumExplicitHs(0)
+            # let UpdatePropertyCache recalculate the number of radicals later
+            atom.SetNumRadicalElectrons(0)
+        # add missing Hs
+        rwmol = Chem.RWMol(new)
+        for atom_idx, hydrogens in reverse_mapping.items():
+            for hydrogen in hydrogens:
+                h_idx = rwmol.AddAtom(hydrogen)
+                rwmol.AddBond(atom_idx, h_idx, Chem.BondType.SINGLE)
+        new = rwmol.GetMol()
+        # sanitize
+        new.UpdatePropertyCache()
+        sanitize_mol(new)
+        # reorder atoms as input atomgroup (through _MDAnalysis_index)
+        return reorder_atoms(new, field=index_field)
+
+
+@dataclass(frozen=True)
+class RDKitInferer:
+    """Uses RDKit's :func:`~rdkit.Chem.rdDetermineBonds.DetermineBondOrders`
+    to infer bond orders and formal charges. This is the same algorithm used
+    by the :ref:`xyz2mol <https://github.com/jensengroup/xyz2mol>` package.
+    """
+
+    charge: int = 0
+
+    def __call__(self, mol: "Chem.Mol") -> "Chem.Mol":
+        new = Chem.Mol(mol)
+        DetermineBondOrders(new, charge=self.charge)
```

Comment:
RDKit now also has a `DetermineBonds` which could be interesting to add as an alternative to `guess_bonds`? Or at least in the RDKitConverter which requires bonds anyway